woocommerce/woocommerce-ios
# Errors: Banner message for error loading data on Product List

Part of: #4382

## Description

To improve our error states, we are introducing a top banner that can be used whenever there is a problem loading data. The banner is collapsed by default and on select, users can read our troubleshooting tips or contact support directly.

This PR sets up the banner and adds it to the Products screen. When there is an error, the error banner is the only banner displayed on the screen. Otherwise, the usual Products banner will be displayed when available. (This banner replaces the "Unable to refresh list" notice we displayed before.)

## Changes

* Sets up the error banner in `ErrorTopBannerFactory`, with two actions (Troubleshoot and Contact support).
* Adds unit tests for the error banner actions in `ErrorTopBannerFactoryTests`.
* Adds the error banner to `ProductsViewController`, and removes the "Unable to refresh list" notice there. Any sync errors will set `hasErrorLoadingData` to `true`, which determines whether the error banner is displayed instead of the usual products top banner.

Collapsed (empty state) | Expanded (empty state) | Expanded (cached data)
-|-|-
![Simulator Screen Shot - iPhone 11 Pro - 2021-06-11 at 14 19 52](https://user-images.githubusercontent.com/8658164/121699178-9eb61300-cac6-11eb-8d9a-db017e48b4ff.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-06-11 at 14 20 08](https://user-images.githubusercontent.com/8658164/121699187-9fe74000-cac6-11eb-8d12-2268c9662463.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-06-11 at 14 14 24](https://user-images.githubusercontent.com/8658164/121699203-a4135d80-cac6-11eb-9024-38fc68ae7633.png)


## Testing

1. Build and run a fresh install of the app (to begin without any cached data).
2. Log in to the app.
3. Disconnect your internet connection. (This is so you will trigger a sync error; you can also disrupt your store's Jetpack connection for a more realistic scenario.)
4. Select Products to open the product list.
5. Notice the error banner appears at the top of the Products empty state screen.
6. Expand the banner and confirm the "Troubleshoot" button opens a troubleshooting doc, and "Contact support" opens a new support ticket in the app.
7. Reconnect to the internet.
8. Pull to refresh the Products list. When your products load, notice the error banner is replaced by the regular products banner.
9. Disconnect from the internet again, pull to refresh, and notice the products banner is replaced by the error banner again.

## Submitter Checklist

Update release notes:
